support package.json#os property

https://docs.npmjs.com/files/package.json#os

try to fix https://github.com/cnpm/cnpm/issues/103

```bash
npm i fsevents
npm ERR! Linux 3.13.0-32-generic
npm ERR! argv "/home/www/.nvm/versions/node/v4.4.1/bin/node" "/home/www/.nvm/versions/node/v4.4.1/bin/npm" "i" "fsevents"
npm ERR! node v4.4.1
npm ERR! npm  v2.14.20
npm ERR! code EBADPLATFORM

npm ERR! notsup Unsupported
npm ERR! notsup Not compatible with your operating system or architecture: fsevents@1.0.12
npm ERR! notsup Valid OS:    darwin
npm ERR! notsup Valid Arch:  any
npm ERR! notsup Actual OS:   linux
npm ERR! notsup Actual Arch: x64

npm ERR! Please include the following file with any support request:
npm ERR!     /tmp/npm-debug.log
```

--- a/.travis.yml
+++ b/.travis.yml
@@ -4,7 +4,6 @@ env:
 language: node_js
 node_js:
   - '6'
-  - '5'
   - '4'
 before_script:
   - echo -e "Host *\n\tStrictHostKeyChecking no\n" >> ~/.ssh/config

--- a/appveyor.yml
+++ b/appveyor.yml
@@ -3,7 +3,7 @@ environment:
     npm_registry: https://registry.npmjs.com
 
   matrix:
-    - nodejs_version: "5"
+    - nodejs_version: "6"
     - nodejs_version: "4"
 
 # Install scripts. (runs after repo cloning)

--- a/bin/install.js
+++ b/bin/install.js
@@ -25,7 +25,6 @@ const fs = require('mz/fs');
 const parseArgs = require('minimist');
 const utils = require('../lib/utils');
 const config = require('../lib/config');
-const get = require('../lib/get');
 const installLocal = require('..').installLocal;
 const installGlobal = require('..').installGlobal;
 
@@ -145,7 +144,7 @@ co(function*() {
   let binaryMirrors = {};
 
   if (inChina) {
-    binaryMirrors = yield getBinaryMirrors(registry);
+    binaryMirrors = yield utils.getBinaryMirrors(registry);
   }
 
   const config = {
@@ -252,35 +251,4 @@ function* updateDependencies(root, pkgs, propName, saveExact) {
   }
   pkg[propName] = newDeps;
   yield fs.writeFile(pkgFile, JSON.stringify(pkg, null, 2));
-}
-
-function* getBinaryMirrors(registry) {
-  const registries = [ registry ].concat([
-    'https://registry.npm.taobao.org',
-    'https://r.cnpmjs.org',
-    'https://registry.npmjs.org',
-  ]);
-  let lastErr;
-  let binaryMirrors;
-  for (const registry of registries) {
-    const binaryMirrorUrl = registry + '/binary-mirror-config/latest';
-    try {
-      const res = yield get(binaryMirrorUrl, {
-        dataType: 'json',
-        followRedirect: true,
-      });
-      binaryMirrors = res.data.mirrors.china;
-      debug('GET %s => %j', binaryMirrorUrl, binaryMirrors);
-      break;
-    } catch (err) {
-      lastErr = err;
-    }
-  }
-
-  if (!binaryMirrors) {
-    console.warn('Get /binary-mirror-config/latest from %s error: %s', registry, lastErr.stack);
-    binaryMirrors = require('binary-mirror-config/package.json').mirrors.china;
-  }
-
-  return binaryMirrors;
 }

--- a/lib/config.js
+++ b/lib/config.js
@@ -19,6 +19,7 @@ const config = {
     npm_config_loglevel: 'http',
   },
   chineseMirrorEnv: {
+    NODEJS_ORG_MIRROR: 'https://npm.taobao.org/mirrors/node',
     NVM_NODEJS_ORG_MIRROR: 'https://npm.taobao.org/mirrors/node',
     NVM_IOJS_ORG_MIRROR: 'https://npm.taobao.org/mirrors/iojs',
     PHANTOMJS_CDNURL: 'https://npm.taobao.org/mirrors/phantomjs',

--- a/lib/install.js
+++ b/lib/install.js
@@ -122,13 +122,40 @@ function* _install(parentDir, pkg, options) {
   }
 
   if (realPkg.deprecated) {
-    const pkgPath = path.join(parentDir, 'node_modules', realPkg.name);
+    const pkgPath = path.join(parentDir, 'node_modules', realPkg.name).replace(options.root, '.');
     options.console.warn('[%s](%s) %s: %s',
       chalk.red(`${realPkg.name}@${realPkg.version}`),
       chalk.grey(pkgPath),
       chalk.magenta('deprecate'),
       realPkg.deprecated
     );
+  }
+
+  // https://docs.npmjs.com/files/package.json#os
+  if (Array.isArray(realPkg.os) && realPkg.os.length > 0) {
+    let match = false;
+    const platform = process.platform;
+    for (const name of realPkg.os) {
+      if (name === platform) {
+        match = true;
+        break;
+      }
+      if (name[0] === '!' && name.substring(1) === platform) {
+        match = false;
+        break;
+      }
+    }
+    if (!match) {
+      const pkgPath = path.join(parentDir, 'node_modules', realPkg.name).replace(options.root, '.');
+      const err = new Error('Package require os(' + realPkg.os.join(', ') +
+        ') not compatible with your platform(' + platform + ')');
+      options.console.error('[%s](%s) %s: %s',
+        chalk.red(`${realPkg.name}@${realPkg.version}`),
+        chalk.grey(pkgPath),
+        chalk.magenta('unsupported'),
+        err.message);
+      throw err;
+    }
   }
 
   // install steps:

--- a/lib/local_install.js
+++ b/lib/local_install.js
@@ -292,7 +292,7 @@ function* runPostInstallTasks(options) {
     try {
       if (installScript) {
         options.console.log(chalk.yellow('[%s@%s] scripts.install: %j at %s'),
-          pkg.name, pkg.version, installScript, root);
+          pkg.name, pkg.version, installScript, root.replace(options.root, '.'));
         const start = Date.now();
         yield utils.runScript(root, installScript, options);
         options.console.log(chalk.yellow('[%s@%s] scripts.install success, use %s'),
@@ -300,7 +300,7 @@ function* runPostInstallTasks(options) {
       }
       if (postinstallScript) {
         options.console.log(chalk.yellow('[%s@%s] scripts.postinstall: %j at %s'),
-          pkg.name, pkg.version, postinstallScript, root);
+          pkg.name, pkg.version, postinstallScript, root.replace(options.root, '.'));
         const start = Date.now();
         yield utils.runScript(root, postinstallScript, options);
         options.console.log(chalk.yellow('[%s@%s] scripts.postinstall success, use %s'),

--- a/lib/utils.js
+++ b/lib/utils.js
@@ -271,3 +271,33 @@ exports.getTarballStream = function* (url, options) {
   }
   return result.res;
 };
+
+exports.getBinaryMirrors = function* getBinaryMirrors(registry) {
+  const registries = [ registry ].concat([
+    'https://registry.npm.taobao.org',
+    'https://r.cnpmjs.org',
+    'https://registry.npmjs.org',
+  ]);
+  let lastErr;
+  let binaryMirrors;
+  for (const registry of registries) {
+    const binaryMirrorUrl = registry + '/binary-mirror-config/latest';
+    try {
+      const res = yield get(binaryMirrorUrl, {
+        dataType: 'json',
+        followRedirect: true,
+      });
+      binaryMirrors = res.data.mirrors.china;
+      break;
+    } catch (err) {
+      lastErr = err;
+    }
+  }
+
+  if (!binaryMirrors) {
+    console.warn('Get /binary-mirror-config/latest from %s error: %s', registry, lastErr.stack);
+    binaryMirrors = require('binary-mirror-config/package.json').mirrors.china;
+  }
+
+  return binaryMirrors;
+};

--- a/test/bundleDependencies.test.js
+++ b/test/bundleDependencies.test.js
@@ -60,18 +60,15 @@ describe('test/bundleDependencies.test.js', function() {
     assert(exists);
   });
 
-  if (process.platform !== 'win32') {
-    // fsevents not support windows
-    it('should link bundleDependencies bin', function*() {
-      yield npminstall({
-        root: tmp,
-        pkgs: [{
-          name: 'fsevents',
-          version: '1.0.6',
-        }],
-      });
-      const bins = yield fs.readdir(path.join(tmp, 'node_modules/fsevents/node_modules/.bin'));
-      assert.deepEqual(bins, ['node-pre-gyp']);
+  it('should link bundleDependencies bin', function*() {
+    yield npminstall({
+      root: tmp,
+      pkgs: [{
+        name: 'sqlite3',
+        version: '3.1.3',
+      }],
     });
-  }
+    const bins = yield fs.readdir(path.join(tmp, 'node_modules/sqlite3/node_modules/.bin'));
+    assert.deepEqual(bins, ['node-pre-gyp']);
+  });
 });

--- a/test/fixtures/specify-darwin-os/index.js
+++ b/test/fixtures/specify-darwin-os/index.js
@@ -1,0 +1,1 @@
+console.log(require('fsevents'));

--- a/test/fixtures/specify-darwin-os/package.json
+++ b/test/fixtures/specify-darwin-os/package.json
@@ -1,0 +1,9 @@
+{
+  "name": "test-os-darwin",
+  "dependencies": {
+    "fsevents": "*"
+  },
+  "scripts": {
+    "postinstall": "node index.js"
+  }
+}

--- a/test/fixtures/specify-linux-os/index.js
+++ b/test/fixtures/specify-linux-os/index.js
@@ -1,0 +1,1 @@
+console.log(require('test-os-linux/package.json'));

--- a/test/fixtures/specify-linux-os/package.json
+++ b/test/fixtures/specify-linux-os/package.json
@@ -1,0 +1,9 @@
+{
+  "name": "test-os-linux",
+  "dependencies": {
+    "test-os-linux": "*"
+  },
+  "scripts": {
+    "postinstall": "node index.js"
+  }
+}

--- a/test/npminstall.js
+++ b/test/npminstall.js
@@ -14,23 +14,25 @@
 
 const npminstall = require('../');
 const config = require('../lib/config');
+const utils = require('../lib/utils');
 
 module.exports = function*(options) {
-  formatOptions(options);
+  yield formatOptions(options);
   return yield npminstall(options);
 };
 
 module.exports.installGlobal = function*(options) {
-  formatOptions(options);
+  yield formatOptions(options);
   return yield npminstall.installGlobal(options);
 };
 
-function formatOptions(options) {
+function* formatOptions(options) {
   if (process.env.local) {
     options.registry = config.chineseRegistry;
     options.env = options.env || {};
     for (const key in config.chineseMirrorEnv) {
       options.env[key] = config.chineseMirrorEnv[key];
     }
+    options.binaryMirrors = yield utils.getBinaryMirrors(options.registry);
   }
 }

--- a/test/specify-os-property.test.js
+++ b/test/specify-os-property.test.js
@@ -1,0 +1,62 @@
+/**
+ * Copyright(c) cnpm and other contributors.
+ * MIT Licensed
+ *
+ * Authors:
+ *   dead_horse <dead_horse@qq.com>
+ */
+
+'use strict';
+
+/**
+ * Module dependencies.
+ */
+
+const assert = require('assert');
+const path = require('path');
+const rimraf = require('rimraf');
+const npminstall = require('./npminstall');
+
+describe('test/specify-os-property.test.js', () => {
+  describe('install fail', () => {
+    const wrongPlatform = process.platform === 'darwin' ? 'linux' : 'darwin';
+    const root = path.join(__dirname, 'fixtures', `specify-${wrongPlatform}-os`);
+
+    function cleanup() {
+      rimraf.sync(path.join(root, 'node_modules'));
+    }
+
+    beforeEach(cleanup);
+    afterEach(cleanup);
+
+    it('should throw error when package.json#os property not match current platform', function*() {
+      try {
+        yield npminstall({
+          root: root,
+        });
+        throw new Error('should not run this');
+      } catch (err) {
+        assert(/not compatible with your platform/.test(err.message), err.message);
+      }
+    });
+  });
+
+  if (process.platform === 'darwin' || process.platform === 'linux') {
+    describe('install success', () => {
+      const root = path.join(__dirname, 'fixtures', `specify-${process.platform}-os`);
+
+      function cleanup() {
+        rimraf.sync(path.join(root, 'node_modules'));
+      }
+
+      beforeEach(cleanup);
+      afterEach(cleanup);
+
+      it('should success when package.json#os property match current platform', function*() {
+        yield npminstall({
+          root: root,
+        });
+      });
+    });
+  }
+});

